Nodejs/Product/Nodejs/Project/OperationProgressService.cs

```diff
@@ -0,0 +1,65 @@
+﻿using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.NodejsTools.Telemetry;
+using Microsoft.VisualStudio.OperationProgress;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Threading;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.NodejsTools.Project
+{
+    internal class OperationProgressService
+    {
+        private readonly Dictionary<string, TaskCompletionSource<bool>> taskCompletionSources = new Dictionary<string, TaskCompletionSource<bool>>();
+
+        public void RegisterAndStartStage(string stageId, string message)
+        {
+            var taskCompletionSource = new TaskCompletionSource<bool>();
+            this.taskCompletionSources.Add(stageId, taskCompletionSource);
+
+            ThreadHelper.JoinableTaskFactory.RunAsync(async () =>
+            {
+                await ThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+
+                if (!(ServiceProvider.GlobalProvider.GetService(typeof(SVsOperationProgress)) is IVsOperationProgress2 operationProgress))
+                {
+                    return;
+                }
+
+                // Register a stage with unique ID and desired display text
+                operationProgress.RegisterStageId(stageId, new OperationProgressStageOptions(message));
+
+                // Register a task to our stage
+                _ = operationProgress.RegisterStageOperationTasksAsync(
```

Comment:
Can you add comments explaining why we're using the API in this way?
For example, something about making the RegisterStateOperationTasks task taking a long time gives us the kind of progress bar we want.

Reply:
Added.